File: scripts/solar/solar-common/actor/util/ActorBuffUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__InstanceOf = ____lualib.__TS__InstanceOf
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 1,["9"] = 1,["10"] = 2,["11"] = 2,["12"] = 3,["13"] = 3,["14"] = 4,["15"] = 4,["16"] = 10,["17"] = 10,["18"] = 10,["20"] = 10,["21"] = 16,["22"] = 17,["23"] = 17,["24"] = 18,["25"] = 19,["27"] = 17,["28"] = 16,["29"] = 27,["30"] = 28,["31"] = 28,["32"] = 29,["33"] = 30,["35"] = 28,["36"] = 27,["37"] = 38,["38"] = 39,["39"] = 39,["40"] = 40,["41"] = 41,["43"] = 39,["44"] = 38,["45"] = 57,["46"] = 58,["47"] = 59,["48"] = 60,["49"] = 61,["51"] = 64,["52"] = 65,["53"] = 66,["54"] = 68,["55"] = 69,["56"] = 70,["57"] = 71,["58"] = 71,["59"] = 71,["60"] = 72,["61"] = 71,["62"] = 71,["66"] = 78,["67"] = 79,["69"] = 82,["70"] = 82,["71"] = 82,["72"] = 82,["73"] = 82,["74"] = 82,["75"] = 82,["76"] = 57,["77"] = 92,["78"] = 93,["80"] = 94,["81"] = 94,["82"] = 95,["83"] = 96,["86"] = 99,["87"] = 100,["88"] = 101,["90"] = 103,["91"] = 94,["94"] = 105,["95"] = 92,["96"] = 113,["97"] = 114,["98"] = 115,["99"] = 116,["100"] = 117,["101"] = 118,["105"] = 122,["106"] = 113,["107"] = 131,["108"] = 132,["109"] = 131,["110"] = 142,["111"] = 143,["112"] = 144,["113"] = 145,["115"] = 142,["116"] = 153,["117"] = 154,["118"] = 155,["119"] = 153,["120"] = 194,["121"] = 195,["122"] = 197,["123"] = 198,["124"] = 199,["125"] = 200,["126"] = 201,["127"] = 202,["131"] = 197,["133"] = 209,["134"] = 210,["135"] = 211,["136"] = 212,["137"] = 213,["140"] = 209,["142"] = 194,["143"] = 224,["144"] = 225,["145"] = 226,["148"] = 229,["149"] = 230,["150"] = 231,["151"] = 232,["152"] = 233,["156"] = 237,["157"] = 224,["158"] = 253,["159"] = 254,["160"] = 255,["163"] = 258,["164"] = 259,["165"] = 260,["166"] = 261,["170"] = 253,["171"] = 273,["172"] = 274,["174"] = 274,["176"] = 273,["177"] = 282,["178"] = 282,["179"] = 282,["181"] = 283,["182"] = 284,["185"] = 287,["186"] = 288,["187"] = 289,["188"] = 290,["191"] = 293,["192"] = 293,["196"] = 282,["197"] = 304,["198"] = 305,["199"] = 306,["202"] = 309,["203"] = 310,["204"] = 310,["205"] = 311,["208"] = 304});
local ____exports = {}
local ____ActorBuff = require("solar.solar-common.actor.ActorBuff")
local ActorBuff = ____ActorBuff.default
local ____Actor = require("solar.solar-common.actor.Actor")
local Actor = ____Actor.default
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
local ____BaseUtil = require("solar.solar-common.util.BaseUtil")
local BaseUtil = ____BaseUtil.default
____exports.default = __TS__Class()
local ActorBuffUtil = ____exports.default
ActorBuffUtil.name = "ActorBuffUtil"
function ActorBuffUtil.prototype.____constructor(self)
end
function ActorBuffUtil.addAnyActorBuffCreatedListener(self, onActorBuffCreatedListener)
    local ____Actor__sl_anyActorCreatedListeners_0 = Actor._sl_anyActorCreatedListeners
    ____Actor__sl_anyActorCreatedListeners_0[#____Actor__sl_anyActorCreatedListeners_0 + 1] = function(____, buff)
        if __TS__InstanceOf(buff, ActorBuff) then
            onActorBuffCreatedListener(nil, buff)
        end
    end
end
function ActorBuffUtil.addAnyActorBuffLevelChangeListener(self, onActorBuffLevelChangeListener)
    local ____Actor__sl_anyActorLevelChangeListeners_1 = Actor._sl_anyActorLevelChangeListeners
    ____Actor__sl_anyActorLevelChangeListeners_1[#____Actor__sl_anyActorLevelChangeListeners_1 + 1] = function(____, buff, delta)
        if __TS__InstanceOf(buff, ActorBuff) then
            onActorBuffLevelChangeListener(nil, buff, delta)
        end
    end
end
function ActorBuffUtil.addAnyActorBuffDestroyListener(self, onActorBuffDestroyListener)
    local ____Actor__sl_anyActorDestroyListeners_2 = Actor._sl_anyActorDestroyListeners
    ____Actor__sl_anyActorDestroyListeners_2[#____Actor__sl_anyActorDestroyListeners_2 + 1] = function(____, buff)
        if __TS__InstanceOf(buff, ActorBuff) then
            onActorBuffDestroyListener(nil, buff)
        end
    end
end
function ActorBuffUtil.addActorBuff(self, unit, actorTypeId, creator, initActorBuff)
    local actorType = DataBase:getSolarActorType(actorTypeId)
    if not actorType then
        log.errorWithTraceBack("buff不存在!请先使用registerActorType注册演员类型! id=" .. actorTypeId)
        return nil
    end
    local oldUnitBuff = ____exports.default:getUnitActorBuff(unit, actorTypeId)
    if oldUnitBuff then
        if actorType.levelMax and actorType.levelMax ~= 0 then
            if actorType.levelMax == -1 or oldUnitBuff.level < actorType.levelMax then
                oldUnitBuff.level = oldUnitBuff.level + 1
                if actorType.levelDurMode == true then
                    BaseUtil.runLater(
                        actorType.dur,
                        function()
                            oldUnitBuff.level = oldUnitBuff.level - 1
                        end
                    )
                end
            end
        end
        oldUnitBuff:updateDurStartTime()
        return oldUnitBuff
    end
    return __TS__New(
        ActorBuff,
        actorTypeId,
        unit,
        creator,
        initActorBuff
    )
end
function ActorBuffUtil.getPlayerUnitBuff(self, player, buffTypeId)
    GroupEnumUnitsOfPlayer(_sl_tempGroup3, player, nil)
    do
        local i = 0
        while i <= 1000000 do
            local unitHandle = FirstOfGroup(_sl_tempGroup3)
            if not IsHandle(unitHandle) then
                break
            end
            local buff = ____exports.default:getUnitActorBuff(unitHandle, buffTypeId)
            if (buff and buff.buffType.id) == buffTypeId then
                return buff
            end
            GroupRemoveUnit(_sl_tempGroup3, unitHandle)
            i = i + 1
        end
    end
    return nil
end
function ActorBuffUtil.getUnitActorBuff(self, unit, actorTypeId)
    local actorBuffs = ____exports.default:getUnitActorBuffs(unit)
    if actorBuffs then
        for ____, actorBuff in ipairs(actorBuffs) do
            if actorBuff.actorTypeId == actorTypeId then
                return actorBuff
            end
        end
    end
    return nil
end
function ActorBuffUtil.isUnitHasActorBuff(self, unit, actorTypeId)
    return ____exports.default:getUnitActorBuff(unit, actorTypeId) ~= nil
end
function ActorBuffUtil.ifUnitHasActorBuff(self, unit, callBack, actorTypeId)
    local actorBuff = ____exports.default:getUnitActorBuff(unit, actorTypeId)
    if actorBuff then
        callBack(nil, actorBuff)
    end
end
function ActorBuffUtil.getUnitActorBuffs(self, unit)
    local unitSolarData = DataBase:getUnitSolarData(unit, false)
    return unitSolarData and unitSolarData._SL_solarActorBuffs
end
function ActorBuffUtil.forAllActorBuffs(self, callback, actorBuffTypeId)
    if actorBuffTypeId then
        DataBase:forUnitSolarDatas(function(____, id, solarData)
            local _SL_solarActorBuffSet = solarData and solarData._SL_solarActorBuffs
            if _SL_solarActorBuffSet then
                for ____, actorBuff in ipairs(_SL_solarActorBuffSet) do
                    if (actorBuff and actorBuff.actorType.id) == actorBuffTypeId then
                        callback(nil, actorBuff)
                    end
                end
            end
        end)
    else
        DataBase:forUnitSolarDatas(function(____, id, solarData)
            local _SL_solarActorBuffSet = solarData and solarData._SL_solarActorBuffs
            if _SL_solarActorBuffSet then
                for ____, actorBuff in ipairs(_SL_solarActorBuffSet) do
                    callback(nil, actorBuff)
                end
            end
        end)
    end
end
function ActorBuffUtil.getUnitActorBuffsByClass(self, unit, clazz)
    local actorBuffSet = ____exports.default:getUnitActorBuffs(unit)
    if not actorBuffSet then
        return
    end
    local result = {}
    for ____, actorBuff in ipairs(actorBuffSet) do
        if actorBuff then
            if clazz == nil or clazz == actorBuff:get("class") then
                result[#result + 1] = actorBuff
            end
        end
    end
    return result
end
function ActorBuffUtil.forUnitActorBuffs(self, unit, callback, clazz)
    local actorBuffSet = ____exports.default:getUnitActorBuffs(unit)
    if not actorBuffSet then
        return
    end
    for ____, actorBuff in ipairs(actorBuffSet) do
        if actorBuff then
            if clazz == nil or clazz == actorBuff:get("class") then
                callback(nil, actorBuff)
            end
        end
    end
end
function ActorBuffUtil.clearUnitActorBuff(self, unit, actorBuffTypeId)
    local ____opt_13 = ____exports.default:getUnitActorBuff(unit, actorBuffTypeId)
    if ____opt_13 ~= nil then
        ____opt_13:destroy()
    end
end
function ActorBuffUtil.clearUnitActorBuffs(self, unit, onlyDeBuff)
    if onlyDeBuff == nil then
        onlyDeBuff = false
    end
    local actorBuffs = ____exports.default:getUnitActorBuffs(unit)
    if not actorBuffs then
        return
    end
    for ____, actorBuff in ipairs(actorBuffs) do
        if onlyDeBuff then
            if actorBuff and actorBuff.actorType.isDebuff then
                actorBuff:destroy()
            end
        else
            if actorBuff ~= nil then
                actorBuff:destroy()
            end
        end
    end
end
function ActorBuffUtil.clearUnitActorBuffsByClass(self, unit, clazz)
    local actorBuffs = ____exports.default:getUnitActorBuffs(unit)
    if not actorBuffs then
        return
    end
    for ____, actorBuff in ipairs(actorBuffs) do
        local ____opt_19 = actorBuff and actorBuff.actorType
        if (____opt_19 and ____opt_19.class) == clazz then
            actorBuff:destroy()
        end
    end
end
return ____exports

File: src/solar/solar-common/actor/util/ActorBuffUtil.ts
import ActorBuff from "@/ActorBuff";
import Actor from "@/Actor";
import DataBase from "@/DataBase";
import BaseUtil from "@/BaseUtil";


/**
 * see ActorUtil 优先使用演员工具 不要把演员局限到技能或物品或buff上 这样会导致演员不够通用切换形态
 */
export default class ActorBuffUtil {


    /**
     * 添加任意演员Buff创建事件 监听回调
     */
    static addAnyActorBuffCreatedListener(onActorBuffCreatedListener: (buff: ActorBuff) => void) {
        Actor._sl_anyActorCreatedListeners.push(buff => {
            if (buff instanceof ActorBuff) {
                onActorBuffCreatedListener(buff);
            }
        });
    }

    /**
     * 添加任意演员Buff叠加层数改变事件 监听回调
     */
    static addAnyActorBuffLevelChangeListener(onActorBuffLevelChangeListener: (buff: ActorBuff, delta: number) => void) {
        Actor._sl_anyActorLevelChangeListeners.push((buff, delta) => {
            if (buff instanceof ActorBuff) {
                onActorBuffLevelChangeListener(buff, delta);
            }
        });
    }

    /**
     * 添加任意演员Buff销毁事件 监听回调
     */
    static addAnyActorBuffDestroyListener(onActorBuffDestroyListener: (buff: ActorBuff) => void) {
        Actor._sl_anyActorDestroyListeners.push(buff => {
            if (buff instanceof ActorBuff) {
                onActorBuffDestroyListener(buff);
            }
        });
    }

    /**
     * 演员buff
     */

    /**
     * 给单位上一个演员buff
     * @param unit 给哪个单位上
     * @param actorTypeId 演员类型id
     * @param creator buff来源单位
     * @param initActorBuff 初始化ActorBuff回调 会在  buffType.onCreated 之前调用  以给buff设置初始自定义值
     */
    static addActorBuff(unit: unit, actorTypeId: string, creator?: unit, initActorBuff?: (actorBuff: ActorBuff) => void): ActorBuff {
        let actorType: AppActorBuffType = DataBase.getSolarActorType(actorTypeId);
        if (!actorType) {
            log.errorWithTraceBack("buff不存在!请先使用registerActorType注册演员类型! id=" + actorTypeId)
            return null;
        }
        //判断是否已存在buff
        let oldUnitBuff = ActorBuffUtil.getUnitActorBuff(unit, actorTypeId);
        if (oldUnitBuff) {
            if (actorType.levelMax && actorType.levelMax != 0) {//可叠加
                //-1 = 无限叠加
                if (actorType.levelMax == -1 || oldUnitBuff.level < actorType.levelMax) {
                    oldUnitBuff.level++;//增加叠加层数
                    if (actorType.levelDurMode == true) {
                        BaseUtil.runLater(actorType.dur, () => {
                            oldUnitBuff.level--;
                        })
                    }

                }
            }
            oldUnitBuff.updateDurStartTime()
            return oldUnitBuff;
        }
        //新建buff
        return new ActorBuff(actorTypeId, unit, creator, initActorBuff);
    }


    /**
     * 获取某个玩家的所有单位拥有的增益效果的 第一个
     * （使用场景：比如获取羁绊UI显示的信息）
     * @param player
     * @param buffTypeId
     */
    static getPlayerUnitBuff(player: player, buffTypeId: string): ActorBuff {
        GroupEnumUnitsOfPlayer(_sl_tempGroup3, player, null)
        for (let i = 0; i <= 1000000; i++) {
            let unitHandle = FirstOfGroup(_sl_tempGroup3)
            if (!IsHandle(unitHandle)) {
                break
            }
            let buff = ActorBuffUtil.getUnitActorBuff(unitHandle, buffTypeId);
            if (buff?.buffType.id == buffTypeId) {
                return buff
            }
            GroupRemoveUnit(_sl_tempGroup3, unitHandle)
        }
        return null;
    }

    /**
     * 获取单位的某个类型的buff
     * @param unit
     * @param actorTypeId
     */
    static getUnitActorBuff(unit: unit, actorTypeId: string): ActorBuff {
        let actorBuffs = ActorBuffUtil.getUnitActorBuffs(unit);
        if (actorBuffs) {
            for (let actorBuff of actorBuffs) {
                if (actorBuff.actorTypeId == actorTypeId) {
                    return actorBuff
                }
            }
        }
        return null;
    }

    /**
     * 使用此函数前请优先考虑使用ActorUtil.isUnitHasActor 以尽量保证演员的功能不局限于某一种演员类型 方便吞噬等转换
     * 判断是否拥有某个类型的ActorBuff
     * @param unit
     * @param actorTypeId
     */
    static isUnitHasActorBuff(unit: unit, actorTypeId: string): boolean {
        return ActorBuffUtil.getUnitActorBuff(unit, actorTypeId) != null;
    }

    /**
     * 使用此函数前请优先考虑使用ActorUtil.ifUnitHasActor（那样其他演员物品 技能等也有效，游戏代码写灵活点才能应对随时可能改变的玩法）
     * 如果拥有某个类型的ActorBuff 则执行回调
     * @param unit
     * @param callBack
     * @param actorTypeId
     */
    static ifUnitHasActorBuff(unit: unit, callBack: (actorBuff: ActorBuff) => void, actorTypeId: string): void {
        let actorBuff = ActorBuffUtil.getUnitActorBuff(unit, actorTypeId);
        if (actorBuff) {
            callBack(actorBuff);
        }
    }

    /**
     * 获取单位身上所有ActorBuff
     * @param unit
     */
    static getUnitActorBuffs(unit: unit): ActorBuff[] {
        let unitSolarData = DataBase.getUnitSolarData(unit, false);
        return unitSolarData?._SL_solarActorBuffs;
    }


    // /**
    //  * 将buff集合 根据创建时间排序为buff数组
    //  * @param actorBuffSet
    //  * @param filter 只对需要的buff进行筛选过滤  返回true表示此buff添加到排序结果中 如不传filter回调 则全部添加到排序结果
    //  */
    // static sortByCreateTime(actorBuffSet: ActorBuffSet, filter?: (actorBuff: ActorBuff) => boolean): ActorBuff[] {
    //     if (actorBuffSet == null) {
    //         return null;
    //     }
    //     let actorBuffList: ActorBuff[] = [];
    //     if (filter == null) {
    //         for (let actorBuffSetKey in actorBuffSet) {
    //             actorBuffList.push(actorBuffSet[actorBuffSetKey])
    //         }
    //     } else {
    //         for (let actorBuffSetKey in actorBuffSet) {
    //             let actorBuff = actorBuffSet[actorBuffSetKey];
    //             if (filter(actorBuff)) {
    //                 actorBuffList.push(actorBuff);
    //             }
    //         }
    //     }
    //
    //     actorBuffList.sort((a, b) => {
    //         return a.creationTime - b.creationTime;
    //     });
    //     return actorBuffList;
    // }


    /**
     * 遍历地图上所有单位的所有ActorBuffs
     * @param callback
     * @param actorBuffTypeId 只对某个ActorBuffType做遍历
     */
    static forAllActorBuffs(callback: (ActorBuff: ActorBuff) => void, actorBuffTypeId?: string): void {
        if (actorBuffTypeId) {//遍历单个类型
            //遍历所有
            DataBase.forUnitSolarDatas((id, solarData: AppUnitData) => {
                let _SL_solarActorBuffSet: ActorBuff[] = solarData?._SL_solarActorBuffs;
                if (_SL_solarActorBuffSet) {
                    for (let actorBuff of _SL_solarActorBuffSet) {
                        if (actorBuff?.actorType.id == actorBuffTypeId) {
                            callback(actorBuff)
                        }
                    }
                }
            })
        } else {
            //遍历所有
            DataBase.forUnitSolarDatas((id, solarData: AppUnitData) => {
                let _SL_solarActorBuffSet: ActorBuff[] = solarData?._SL_solarActorBuffs;
                if (_SL_solarActorBuffSet) {
                    for (let actorBuff of _SL_solarActorBuffSet) {
                        callback(actorBuff)
                    }
                }
            })
        }

    }

    /**
     * 获取此单位的 指定Class的所有ActorBuff
     */
    static getUnitActorBuffsByClass(unit: unit, clazz: string): ActorBuff[] {
        let actorBuffSet = ActorBuffUtil.getUnitActorBuffs(unit);
        if (!actorBuffSet) {
            return
        }
        let result: ActorBuff[] = []
        for (let actorBuff of actorBuffSet) {
            if (actorBuff) {
                if (clazz == null || clazz == actorBuff.get("class")) {
                    result.push(actorBuff);
                }
            }
        }
        return result;
    }

    // /**
    //  * 获取此单位的  所有ActorBuff 并按创建时间排序
    //  * @param unit
    //  * @param filter 返回true 表示此buff添加到排序结果中 如不传filter回调 则全部添加到排序结果
    //  */
    // static getUnitActorBuffsAndSortCreateTime(unit: unit, filter?: (actorBuff: ActorBuff) => boolean): ActorBuff[] {
    //     let actorBuffSet = ActorBuffUtil.getUnitActorBuffs(unit);
    //     return ActorBuffUtil.sortByCreateTime(actorBuffSet, filter);
    // }

    /**
     * 遍历此单位的 所有ActorBuff
     */
    static forUnitActorBuffs(unit: unit, callback: (actorBuff: ActorBuff) => void, clazz?: string): void {
        let actorBuffSet = ActorBuffUtil.getUnitActorBuffs(unit);
        if (!actorBuffSet) {
            return
        }
        for (let actorBuff of actorBuffSet) {
            if (actorBuff) {
                if (clazz == null || clazz == actorBuff.get("class")) {
                    callback(actorBuff);
                }
            }
        }

    }

    /**
     * 清除单位某个类型的ActorBuff
     * @param unit
     * @param actorBuffTypeId
     */
    static clearUnitActorBuff(unit: unit, actorBuffTypeId: string) {
        ActorBuffUtil.getUnitActorBuff(unit, actorBuffTypeId)?.destroy();
    }

    /**
     * 清除单位所有ActorBuff
     * @param unit 单位
     * @param onlyDeBuff 只销毁减益演员buff
     */
    static clearUnitActorBuffs(unit: unit, onlyDeBuff = false) {
        let actorBuffs: ActorBuff[] = ActorBuffUtil.getUnitActorBuffs(unit);
        if (!actorBuffs) {
            return;
        }
        for (let actorBuff of actorBuffs) {
            if (onlyDeBuff) {
                if (actorBuff?.actorType.isDebuff) {
                    actorBuff.destroy();
                }
            } else {
                actorBuff?.destroy();
            }

        }
    }

    /**
     * 清除单位身上某个类别的所有ActorBuff
     * @param unit 单位
     * @param clazz 同 Class
     */
    static clearUnitActorBuffsByClass(unit: unit, clazz: string) {
        let actorBuffs: ActorBuff[] = ActorBuffUtil.getUnitActorBuffs(unit);
        if (!actorBuffs) {
            return;
        }
        for (let actorBuff of actorBuffs) {
            if (actorBuff?.actorType?.class == clazz) {
                actorBuff.destroy()
            }
        }
    }


}